Source/utils/cel_to_cl2.cpp

```diff
@@ -0,0 +1,208 @@
+#include <iostream>
+#endif
+
+#include "appfat.h"
+#include "utils/endian.hpp"
+
+namespace devilution {
+
+namespace {
+
+constexpr bool IsCelTransparent(uint8_t control)
+{
+	constexpr uint8_t CelTransparentMin = 0x80;
+	return control >= CelTransparentMin;
+}
+
+constexpr uint8_t GetCelTransparentWidth(uint8_t control)
+{
+	return -static_cast<int8_t>(control);
+}
+
+void AppendCl2TransparentRun(unsigned width, std::vector<uint8_t> &out)
+{
+	while (width >= 0x7F) {
+		out.push_back(0x7F);
+		width -= 0x7F;
+	}
+	if (width == 0)
+		return;
+	out.push_back(width);
+}
+
+void AppendCl2FillRun(uint8_t color, unsigned width, std::vector<uint8_t> &out)
+{
+	while (width >= 0x3F) {
+		out.push_back(0x80);
+		out.push_back(color);
+		width -= 0x3F;
+	}
+	if (width == 0)
+		return;
+	out.push_back(0xBF - width);
+	out.push_back(color);
+}
+
+void AppendCl2PixelsRun(const uint8_t *src, unsigned width, std::vector<uint8_t> &out)
+{
+	while (width >= 0x41) {
+		out.push_back(0xBF);
+		for (size_t i = 0; i < 0x41; ++i)
+			out.push_back(src[i]);
+		width -= 0x41;
+		src += 0x41;
+	}
+	if (width == 0)
+		return;
+	out.push_back(256 - width);
+	for (size_t i = 0; i < width; ++i)
+		out.push_back(src[i]);
+}
+
+void AppendCl2PixelsOrFillRun(const uint8_t *src, unsigned length, std::vector<uint8_t> &out)
+{
+	const uint8_t *begin = src;
+	const uint8_t *prevColorBegin = src;
+	unsigned prevColorRunLength = 1;
+	uint8_t prevColor = *src++;
+	while (--length > 0) {
+		const uint8_t color = *src;
+		if (prevColor == color) {
+			++prevColorRunLength;
+		} else {
+			// A tunable parameter that decides at which minimum length we encode a fill run.
+			// 3 appears to be optimal for most of our data (much better than 2, rarely very slightly worse than 4).
+			constexpr unsigned MinFillRunLength = 3;
+			if (prevColorRunLength >= MinFillRunLength) {
+				AppendCl2PixelsRun(begin, prevColorBegin - begin, out);
+				AppendCl2FillRun(prevColor, prevColorRunLength, out);
+				begin = src;
+			}
+			prevColorBegin = src;
+			prevColorRunLength = 1;
+			prevColor = color;
+		}
+		++src;
+	}
+	AppendCl2PixelsRun(begin, prevColorBegin - begin, out);
+	AppendCl2FillRun(prevColor, prevColorRunLength, out);
+}
+
+} // namespace
+
+OwnedCelSprite CelToCl2(const uint8_t *data, size_t size, PointerOrValue<uint16_t> widthOrWidths)
+{
+	// A CEL file either begins with:
+	// 1. A CEL header.
+	// 2. A list of offsets to frame groups (each group is a CEL file).
+	size_t groupsHeaderSize = 0;
+	uint32_t numGroups = 1;
+	const uint32_t maybeNumFrames = LoadLE32(data);
+
+	std::vector<uint8_t> cl2Data;
+
+	// Most files become smaller with CL2. Allocate exactly enough bytes to avoid reallocation.
+	// The only file that becomes larger is Data\hf_logo3.CEL, by exactly 4445 bytes.
+	cl2Data.reserve(size + 4445);
+
+	// If it is a number of frames, then the last frame offset will be equal to the size of the file.
+	if (LoadLE32(&data[maybeNumFrames * 4 + 4]) != size) {

```

Comment:
nice to have logic handle this rather then having it hardcoded for every sprite.